### ISSUE TITLE
CTC loss计算这里有一个疑问

### ISSUE THREAD

**Issue by zuxinqi** [wntg/LLaMA-Omni#16]
很棒的项目，感谢大佬，这里有个疑问，
loss = self.speech_generator(llama_output['hidden_states'][-1], labels, tgt_units)
ctc loss计算这里，传入的labels是question的内容，没有answer的传入，tgt_units是answer的tts后的units，这里是不匹配的，这里是不是应该和stage1类似，input ids和labels传入question+answer的内容(这里的answer就是转成tgt_units的那个answer)，labels里面question部分的token为IGNORE_INDEX

**Comment by wntg** [comment]
answer相当于是llama_output['hidden_states'][-1]

**Comment by zuxinqi** [comment]
> answer相当于是llama_output['hidden_states'][-1]

llama_output['hidden_states'][-1]是最后一层的hidden_states，在训练阶段，它只会按照给定的labels去计算loss，现在的labels只有question，所以就是question的最后一层的hidden_states，相当训练的时候llm推理的内容完全没有answer的内容，这样和tgt_units是对应不起来的。

**Comment by zuxinqi** [comment]
> 很棒的项目，感谢大佬，这里有个疑问， loss = self.speech_generator(llama_output['hidden_states'][-1], labels, tgt_units) ctc loss计算这里，传入的labels是question的内容，没有answer的传入，tgt_units是answer的tts后的units，这里是不匹配的，这里是不是应该和stage1类似，input ids和labels传入question+answer的内容(这里的answer就是转成tgt_units的那个answer)，labels里面question部分的token为IGNORE_INDEX

我在项目上进行了上述修改，现在loss可以降到0.1以下了，在训练数据上过拟合了，推理的时候可以输出正常语音，不过由于数据太少，没有泛化性，我提交了一个pr

**Comment by wblgers** [comment]
> 我在项目上进行了上述修改，现在loss可以降到0.1以下了，在训练数据上过拟合了，推理的时候可以输出正常语音，不过由于数据太少，没有泛化性，我提交了一个pr

这个地方说法不对，**现在的labels只有question，所以就是question的最后一层的hidden_states**，现在用的是question输入，的对应的预测输出的最后一层的hidden。确实是应该加上answer的label，并且把前面的设置为IGNORE_INDEX。
很感谢你的修改，我也测试一下你的新代码。

**Comment by jellyfishovo** [comment]
> > 很棒的项目，感谢大佬，这里有个疑问， loss = self.speech_generator(llama_output['hidden_states'][-1], labels, tgt_units) ctc loss计算这里，传入的labels是question的内容，没有answer的传入，tgt_units是answer的tts后的units，这里是不匹配的，这里是不是应该和stage1类似，input ids和labels传入question+answer的内容(这里的answer就是转成tgt_units的那个answer)，labels里面question部分的token为IGNORE_INDEX
> 
> 我在项目上进行了上述修改，现在loss可以降到0.1以下了，在训练数据上过拟合了，推理的时候可以输出正常语音，不过由于数据太少，没有泛化性，我提交了一个pr

想请问一下，我采用了您的方案，按照更新后的代码重新训练，我用的1B模型，在50个epoch后损失在0.5左右，我在实际测试的时候发现，相比较先前的代码。您的方案训练出来的模型确实在音频回复上的时长明显增加，但是回复的语音还是非常不清楚（语速快，只有零星几个单词能听懂），请问您提到的“推理的时候可以输出正常语音”是完全清晰且符合问题的语音吗？我的训练样本还是100个，不知道是不是因为样本量过少导致我的回答不清楚

**Comment by jellyfishovo** [comment]
> > 我在项目上进行了上述修改，现在loss可以降到0.1以下了，在训练数据上过拟合了，推理的时候可以输出正常语音，不过由于数据太少，没有泛化性，我提交了一个pr
> 
> 这个地方说法不对，**现在的labels只有question，所以就是question的最后一层的hidden_states**，现在用的是question输入，的对应的预测输出的最后一层的hidden。确实是应该加上answer的label，并且把前面的设置为IGNORE_INDEX。 很感谢你的修改，我也测试一下你的新代码。

请问您的代码是否能生成正常流畅清晰的语音回复呢，我采用了同样的代码训练，但是结果只是生成的音频在时长上有所增加，质量上并没有很明显的提升

**Comment by wblgers** [comment]
> 训练

英文的可以出很清洗的声音了

**Comment by jellyfishovo** [comment]
> > 训练
> 
> 英文的可以出很清洗的声音了

想请问您训练了多少个epoch呢？我用更新后的代码，LLM用的是3.2-1B模型，在第二阶段跑了50个epoch,效果不太理想，您可否发一份您的训练配置文件让我学习一下？谢谢！

**Comment by ranck626** [comment]
@zuxinqi 请问stage2，为什么要同时输入question和answer呢？
我的理解是
**question**--->llm--->hidden_state--->文本--->TTS音频answer--->**音频Unit**
由于音频unit是根据第一阶段的文本，回复构造的。
那么question和音频unit已经匹配啦

**Comment by ZYJGO** [comment]
@jellyfishovo 请问你后来又增大数据集训练了吗，最后效果怎么样？

**Comment by ZYJGO** [comment]
> 我在项目上进行了上述修改，现在loss可以降到0.1以下了，在训练数据上过拟合了，推理的时候可以输出正常语音，不过由于数据太少，没有泛化性，我提交了一个pr

大约5w条数据训练以后有很清楚的声音了
